rafthttp/functional_test.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2015 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rafthttp
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/net/context"
+	"github.com/coreos/etcd/etcdserver/stats"
+	"github.com/coreos/etcd/pkg/types"
+	"github.com/coreos/etcd/raft/raftpb"
+)
+
+func TestSendMessage(t *testing.T) {
+	recvc := make(chan raftpb.Message, 1)
+	p := &fakeProcessor{recvc: recvc}
+	ss := &stats.ServerStats{}
+	ss.Initialize()
+	tr := NewTransporter(&http.Transport{}, types.ID(1), types.ID(1), p, nil, ss, stats.NewLeaderStats("1"))
+	srv := httptest.NewServer(tr.Handler())
+	defer srv.Close()
+	tr.AddPeer(types.ID(1), []string{srv.URL})
+	defer tr.Stop()
+	waitStreamWorking(tr.(*transport).Peer(types.ID(1)))
+
+	data := []byte("some data")
+	tests := []raftpb.Message{
+		{Type: raftpb.MsgProp, To: 1, Entries: []raftpb.Entry{{Data: data}}},
+		// TODO: send MsgApp out when it fits msgapp stream but the term doesn't match
+		{Type: raftpb.MsgApp, To: 1, Term: 1},
+		{Type: raftpb.MsgAppResp, To: 1, Term: 1},
+		{Type: raftpb.MsgVote, To: 1, Term: 1},
+		{Type: raftpb.MsgVoteResp, To: 1, Term: 1},
+		{Type: raftpb.MsgSnap, To: 1, Term: 1, Snapshot: raftpb.Snapshot{Data: data}},
+		{Type: raftpb.MsgHeartbeat, To: 1, Term: 1},
+		{Type: raftpb.MsgHeartbeatResp, To: 1, Term: 1},
+	}
+	for i, tt := range tests {
+		tr.Send([]raftpb.Message{tt})
+		msg := <-recvc
+		if !reflect.DeepEqual(msg, tt) {
+			t.Errorf("#%d: msg = %+v, want %+v", i, msg, tt)
+		}
+	}
+}
+
+func TestSendMessageWhenStreamBreakage(t *testing.T) {
+	recvc := make(chan raftpb.Message, 1)
+	p := &fakeProcessor{recvc: recvc}
+	ss := &stats.ServerStats{}
+	ss.Initialize()
+	tr := NewTransporter(&http.Transport{}, types.ID(1), types.ID(1), p, nil, ss, stats.NewLeaderStats("1"))
+	srv := httptest.NewServer(tr.Handler())
+	defer srv.Close()
+	tr.AddPeer(types.ID(1), []string{srv.URL})
+	defer tr.Stop()
+	peer := tr.(*transport).Peer(types.ID(1))
+	waitStreamWorking(peer)
+
+	// make stream readers point to non-existant endpoints, so stream
+	// is under breakage and cannot be used any more
+	peer.msgAppReader.update("")
+	peer.reader.update("")
+	var n int
+	for {
+		select {
+		case <-time.After(time.Millisecond):
+			n++
+			tr.Send([]raftpb.Message{{Type: raftpb.MsgProp, To: 1}})
```

Comment:
I want to use a message that can be sent out here, because local message (which empty message is) cannot be sent. I will update it to a real raft message if it is not.